### ISSUE TITLE
release: audit unit/integration upgrade tests

### ISSUE THREAD

**Issue by srosenberg** [cockroachdb/cockroach#100552]
All unit tests which invoke `MakeTestingClusterSettingsWithVersions` _and_ pass a non-default `binaryMinSupportedVersion` (i.e., anything other than `clusterversion.TestingBinaryMinSupportedVersion`) should be audited. The reason is explained in detail in [1].

**TL;DR** Before [1], the test cluster would be bootstrapped at the _current_ version unless the test passed `BootstrapVersionKeyOverride`. Thus, migrations weren't being tested other than their idempotency. After [1], the test cluster is correctly bootstrapped at the default `binaryMinSupportedVersion` unless the test overrides it. This roughly translates to,

* Node is bootstrapped and joins the test cluster at the default `binaryMinSupportedVersion`
* StartServer uses `s.BinaryVersionOverride()` to `SET CLUSTER SETTING version` which triggers migrations from the bootstrapped version to the (overridden) `BinaryVersion`

Below are all unit tests, grouped by package, which should be audited. This set is conservative. The ones which say `FAIL` pass a non-default `binaryMinSupportedVersion`, but may still be correct. The ones which don't say `FAIL` are likely to be correct, but it wouldn't hurt to double-check them.

pkg_ccl_kvccl_kvtenantccl
===========================
TestTenantUpgradeFailure
TestTenantUpgradeInterlock

pkg_ccl_serverccl
==================
TestServerStartupGuardrails (`FAIL`)
TestBumpTenantClusterVersion (`FAIL`)
TestValidateTargetTenantClusterVersion (`FAIL`)

pkg_kv_kvclient_kvcoord
======================
TestBiDirectionalRangefeedNotUsedUntilUpgradeFinalilzed (`FAIL`)

pkg_kv_kvserver
=================
TestMigrateWaitsForApplication (`FAIL`)
TestLeaseUpgradeVersionGate   (`FAIL`)
TestRangeMigration (`FAIL`)
TestLoadBasedRebalancingObjective
TestRebalanceObjectiveManager
TestStoreConfig

pkg_kv_kvserver_batcheval
=========================
TestDeclareKeysResolveIntent

pkg_server_settingswatcher
========================
TestVersionGuard

pkg_storage
==============
TestPebbleIterator_ExternalCorruption

pkg_upgrade_upgrademanager
==========================
TestAlreadyRunningJobsAreHandledProperly (`FAIL`)
TestConcurrentMigrationAttempts (`FAIL`)
TestMigrateUpdatesReplicaVersion (`FAIL`)
TestPauseMigration (`FAIL`)

pkg_upgrade_upgrades
=======================
TestDatabaseRoleSettingsUserIDMigration1500Users (`FAIL`)
TestDeleteDescriptorsOfDroppedFunctions (`FAIL`)
TestExternalConnectionsUserIDMigration10Users (`FAIL`)
TestFixUserfileRelatedDescriptorCorruptionUpgrade (`FAIL`)
TestPreconditionBeforeStartingAnUpgrade (`FAIL`)
TestMigrationWithFailuresMultipleAltersOnSameColumn
TestSystemJobInfoMigration
TestSystemPrivilegesIndexMigration
TestSystemActivityMigration
TestWaitForDelRangeInGCJob (`FAIL`)
TestWebSessionsUserIDMigrationNoUsers

[1] https://github.com/cockroachdb/cockroach/pull/99082

**Comment by srosenberg** [comment]
@cockroachdb/multi-tenant, @cockroachdb/sql-schema, @cockroachdb/storage, @cockroachdb/kv-prs, @cockroachdb/server-prs

**Comment by rafiss** [comment]
`TestDatabaseRoleSettingsUserIDMigration1500Users` is tagged, but it does not call `MakeTestingClusterSettingsWithVersions`. It does set this though:
```
				Server: &server.TestingKnobs{
					DisableAutomaticVersionUpgrade: make(chan struct{}),
					BootstrapVersionKeyOverride:    clusterversion.V22_2,
					BinaryVersionOverride:          clusterversion.ByKey(clusterversion.V23_1DatabaseRoleSettingsHasRoleIDColumn - 1),
				},
```

But so does `TestSystemPrivilegesUserIDMigration1500Users`, and that's not in this list.

**Comment by srosenberg** [comment]
> `TestDatabaseRoleSettingsUserIDMigration1500Users` is tagged, but it does not call `MakeTestingClusterSettingsWithVersions`. 

It used before this change [1]. When I ran this last night, my local `release-23.1` was a few days behind.

[1] https://github.com/cockroachdb/cockroach/commit/16ff79d7eede84679b0fef0025f46313a2b00475#diff-581dd19761966c928337b36[…]174592c84a4c44a699816111b8c5

**Comment by aliher1911** [comment]
`TestRangeMigration` - is ok, it is using made up versions to verify that below raft migration infrastructure
`TestLeaseUpgradeVersionGate` - is ok as it only needs version to test version gates
`TestMigrateWaitsForApplication` - is ok as it is testing migration machinery, not actions.
`TestBiDirectionalRangefeedNotUsedUntilUpgradeFinalilzed` - is ok, we only need version gate regardless of data.

What I noticed when checking tests is that docs on `BinaryVersionOverride` explain intricacies of forcing versions settings and bootstrap override while it would be more helpful to explain how to achieve certain goal. e.g. start test cluster with data at old version to test migration, how to start cluster to validate stateless version gate without paying price of actual migration etc.
I think audited tests could have BinaryVersionOverride removed without any ill effects, but I'm not 100% sure and I'm confident they currently test what's expected.

**Comment by msbutler** [comment]
@srosenberg @adityamaru @rafiss  I spent a bit of time trawling though [this doc](https://docs.google.com/document/d/1agLfont05T0BNwBo2pSP_ZHlg0WXo_C6gd-zOY9fPxY/edit) explaining the problem and the[ lengthy comment](https://github.com/cockroachdb/cockroach/blob/0a0985ec51422c3679b4b1f072b6f3d8ae3f6182/pkg/server/testing_knobs.go#L63) on `testingKnobs.BinaryVersionOverride`, and I'm still not totally clear what the best practice is for mixed version unit test writers. Here's what _I guess_ we ought initialize a test cluster:
```
tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
		ServerArgs: base.TestServerArgs{
			Knobs: base.TestingKnobs{
				Server: &server.TestingKnobs{
					DisableAutomaticVersionUpgrade: make(chan struct{}),
					BinaryVersionOverride:          clusterversion.TestingBinaryMinSupportedVersion,
					BootstrapVersionKeyOverride:    clusterversion.BinaryMinSupportedVersionKey,
				},
			},
		},
	})
```

In other words, a unit test writer needs to specify _both_ the binaryVersionOverride and the BootstrapVersionKeyOverride. I'm inferring this from @adityamaru 's TODO: 
https://github.com/cockroachdb/cockroach/blob/0a0985ec51422c3679b4b1f072b6f3d8ae3f6182/pkg/server/testing_knobs.go#L79-L81

Could somebody clarify and then update the issue header?

**Comment by kvoli** [comment]
`TestLoadBasedRebalancingObjective` only uses the version (v22.2) to test the version gate as one of the sub-tests.
`TestRebalanceObjectiveManager` ditto.

**Comment by srosenberg** [comment]
@msbutler Responding to your comment below...

> In other words, a unit test writer needs to specify both the binaryVersionOverride and the BootstrapVersionKeyOverride.

Exactly! The current logic [1] adds a safety net and conservatively bootstraps at `BinaryMinSupportedVersionKey` in cases where `BootstrapVersionKeyOverride` is omitted _and_ `binaryMinSupportedVersion` was not overridden. For auditing purposes, it suffices to verify that in those cases, the test is doing the "right thing" owing to the fact that the cluster will be bootstrapped at the _current_ version.

[1] https://github.com/cockroachdb/cockroach/blob/530100fd39cc722bc324bfb3869a325622258fb3/pkg/server/node.go#L363-L371

**Comment by renatolabs** [comment]
Trying to find out what is left to be done here.

@cockroachdb/multi-tenant Have you had the chance to take a look at the tests under `pkg/ccl/kvccl/kvtenantccl`, `pkg/ccl/serverccl`, and `pkg/server/settingswatcher`?

@cockroachdb/storage have you had a chance to look at `TestPebbleIterator_ExternalCorruption`?

**Comment by nicktrav** [comment]
The Storage test was updated last week. I've updated the summary with a link to the patch.

**Comment by knz** [comment]
I checked TestTenantUpgradeInterlock.

For the others we're waiting for @healthy-pod and @ajstorm to be back next week.

**Comment by renatolabs** [comment]
Everything has been addressed, closing the issue. Thanks everyone for your help!
